src/hash_picker.cpp

```diff
@@ -0,0 +1,640 @@
+		for (int i = block_index;; i >>= 1)
+		{
+			int const first_check_index = leafs_index + ((i & 1) ? -leafs_size : leafs_size);
+			bool done = false;
+			for (int j = 0; j < std::min(leafs_size, file_num_blocks - first_check_index); ++j)
+			{
+				if (merkle_tree[first_block_index + first_check_index + j].is_all_zeros())
+				{
+					done = true;
+					break;
+				}
+			}
+			if (done) break;
+			if (i & 1) leafs_index -= leafs_size;
+			leafs_size *= 2;
+			root_index = merkle_get_parent(root_index);
+			// if an inner node is known then its parent must be known too
+			// so if the root is known the next sibling subtree should already
+			// be computed if all of its leafs have valid hashes
+			if (!merkle_tree[root_index].is_all_zeros()) break;
+			TORRENT_ASSERT(root_index != 0);
+			TORRENT_ASSERT(leafs_index >= 0);
+			TORRENT_ASSERT(leafs_size <= merkle_num_leafs(file_num_blocks));
+		}
+
+		TORRENT_ASSERT(leafs_index >= 0);
+		TORRENT_ASSERT(leafs_index < merkle_num_leafs(m_files.file_num_blocks(f)));
+		TORRENT_ASSERT(leafs_index + leafs_size > block_index);
+
+		// if the root node is unknown the hashes cannot be verified yet
+		if (merkle_tree[root_index].is_all_zeros())
+			return set_block_hash_result::unknown();
+
+		// save the root hash because merkle_fill_tree will overwrite it
+		sha256_hash root = merkle_tree[root_index];
+		merkle_fill_tree(merkle_tree, leafs_size, first_block_index + leafs_index);
+
+		if (root != merkle_tree[root_index])
+		{
+			// hash failure, clear all the internal nodes
+			merkle_clear_tree(merkle_tree, leafs_size / 2, merkle_get_parent(first_block_index + leafs_index));
+			merkle_tree[root_index] = root;
+			// If the hash failure was detected within a single piece then report a piece failure
+			// otherwise report unknown. The pieces will be checked once their hashes have been
+			// downloaded.
+			if (leafs_size <= m_files.piece_length() / default_block_size)
+				return set_block_hash_result::piece_hash_failed();
+			else
+				return set_block_hash_result::unknown();
+		}
+		else
+		{
+			std::fill_n(m_hash_verified[f].begin() + leafs_index, std::min(leafs_size, file_num_blocks - leafs_index), true);
+		}
+
+		int const blocks_per_piece = m_files.piece_length() / default_block_size;
+		piece_index_t const file_first_piece(int(m_files.file_offset(f) / m_files.piece_length()));
+		return { int(leafs_index - static_cast<int>(piece - file_first_piece) * blocks_per_piece)
+			, std::min(leafs_size, m_files.file_num_pieces(f) * blocks_per_piece - leafs_index) };
+	}
+
+	void hash_picker::hashes_rejected(hash_request const& req)
+	{
+		if (req.base == m_piece_layer && req.index % 512 == 0)
+		{
+			for (int i = req.index; i < req.index + req.count; i += 512)
+			{
+				m_piece_hash_requested[req.file][i / 512].last_request = min_time();
+				--m_piece_hash_requested[req.file][i / 512].num_requests;
+			}
+		}
+		else if (req.base == 0)
+		{
+			priority_block_request block_req(req.file, req.index);
```

Comment:
I must admit, I don't understand the logic here. Why are hash requests high priority under these conditions?
Is this also related to block-level requests for future streaming support?